ClientSideEncryptionMigration/Azure.Storage.Blobs.EncryptionMigration/DataMigrator.cs

```diff
@@ -0,0 +1,173 @@
+﻿using Azure.Storage.Blobs.EncryptionMigration.Models;
+using Azure.Storage.Blobs.Models;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Storage.Blobs.EncryptionMigration
+{
+    public class DataMigrator<TUploadArgs>
+    {
+        private readonly ClientSideEncryptionDownloader _downloader;
+        private readonly IEncryptionUploader<TUploadArgs> _uploader;
+
+        public DataMigrator(ClientSideEncryptionDownloader downloader, IEncryptionUploader<TUploadArgs> uploader)
+        {
+            _downloader = downloader;
+            _uploader = uploader;
+        }
+
+        /// <summary>
+        /// Attempts to migrate a blob off client-side encryption v1, if it's already encrypted that way.
+        /// </summary>
+        /// <param name="blob">Blob to migrate if encrypted.</param>
+        /// <param name="args">Arguments for upload, specific to the chosen new encryption solution.</param>
+        /// <param name="progressHandler">Progress handler for blob transfer.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>True if the blob was migrated. False if the blob does not exist or is not client-side encrypted.</returns>
+        public async Task<bool> MigrateBlobIfClientsideV1Encrypted(BlobClient blob, TUploadArgs args, IProgress<(BlobMigrationState State, long BytesTransfered)> progressHandler, CancellationToken cancellationToken)
+        {
+            var plaintextHolder = new MemoryStream();
```

Comment:
This seems dangerous? What about big blobs?

Reply:
I suppose this is just a sample and customer should modify it for their scenarios.